test/extensions/stats_sinks/common/statsd/udp_statsd_test.cc

```diff
@@ -32,6 +32,39 @@ class MockWriter : public UdpStatsdSink::Writer {
   MOCK_METHOD(void, write, (const std::string& message));
 };
 
+// Regression test for https://github.com/envoyproxy/envoy/issues/8911
+TEST(UdpOverUdsStatsdSinkTest, InitWithIpAddress) {
+  auto uds_address = std::make_shared<Network::Address::PipeInstance>(
+      TestEnvironment::unixDomainSocketPath("udstest.1.sock"));
+  NiceMock<ThreadLocal::MockInstance> tls_;
+  NiceMock<Stats::MockMetricSnapshot> snapshot;
+  UdpStatsdSink sink(tls_, uds_address, false);
+
+  NiceMock<Stats::MockCounter> counter;
+  counter.name_ = "test_counter";
+  counter.used_ = true;
+  counter.latch_ = 1;
+  snapshot.counters_.push_back({1, counter});
+
+  // Flush before the server is running. This will fail.
+  sink.flush(snapshot);
+
+  // Start the server.
+  // TODO(mattklein123): Right now all sockets are non-blocking. Move this non-blocking
+  // modification back to the abstraction layer so it will work for multiple platforms. Additionally
+  // this uses low level networking calls because our abstractions in this area only work for IP
+  // sockets. Revisit this also.
+  auto io_handle = uds_address->socket(Network::Address::SocketType::Datagram);
+  RELEASE_ASSERT(fcntl(io_handle->fd(), F_SETFL, 0) != -1, "");
+  uds_address->bind(io_handle->fd());
+
+  // Do the flush which should have somewhere to write now.
+  sink.flush(snapshot);
+  Buffer::OwnedImpl receive_buffer;
+  receive_buffer.read(*io_handle, 32);
+  EXPECT_EQ("envoy.test_counter:1|c", receive_buffer.toString());
```

Comment:
would this EXPECT fail if we use connect() and write()?

Reply:
Yeah I checked. With this diff:
```
mklein@localhost:~/Source/envoy$ git diff
diff --git a/source/extensions/stat_sinks/common/statsd/statsd.cc b/source/extensions/stat_sinks/common/statsd/statsd.cc
index 5782935fb..27c7950d5 100644
--- a/source/extensions/stat_sinks/common/statsd/statsd.cc
+++ b/source/extensions/stat_sinks/common/statsd/statsd.cc
@@ -28,13 +28,12 @@ namespace Statsd {
 
 UdpStatsdSink::WriterImpl::WriterImpl(UdpStatsdSink& parent)
     : parent_(parent),
-      io_handle_(parent_.server_address_->socket(Network::Address::SocketType::Datagram)) {}
+      io_handle_(parent_.server_address_->socket(Network::Address::SocketType::Datagram)) {
+  parent_.server_address_->connect(io_handle_->fd());
+}
 
 void UdpStatsdSink::WriterImpl::write(const std::string& message) {
-  // TODO(mattklein123): We can avoid this const_cast pattern by having a constant variant of
-  // RawSlice. This can be fixed elsewhere as well.
-  Buffer::RawSlice slice{const_cast<char*>(message.c_str()), message.size()};
-  Network::Utility::writeToSocket(*io_handle_, &slice, 1, nullptr, *parent_.server_address_);
+  ::send(io_handle_->fd(), message.c_str(), message.size(), MSG_DONTWAIT);
 }
 
 UdpStatsdSink::UdpStatsdSink(ThreadLocal::SlotAllocator& tls,
```

Both connect() fails if the assert is put back, and the test hangs.

Reply:
Thanks for trying it out!